src/OrbitQt/AnnotatingSourceCodeHandler.cpp

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "AnnotatingSourceCodeHandler.h"
+
+#include <chrono>
+#include <filesystem>
+
+#include "CodeReport/AnnotateDisassembly.h"
+#include "CodeReport/AnnotatingLine.h"
+#include "ObjectUtils/ElfFile.h"
+#include "OrbitBase/ImmediateExecutor.h"
+#include "OrbitBase/ReadFileToString.h"
+#include "OrbitBase/Result.h"
+#include "QtUtils/MainThreadExecutorImpl.h"
+#include "SourcePathsMapping/MappingManager.h"
+#include "SourcePathsMappingUI/AskUserForFile.h"
+
+namespace orbit_qt {
+AnnotatingSourceCodeHandler::AnnotatingSourceCodeHandler(
+    const QPointer<orbit_code_viewer::OwningDialog>& dialog,
+    orbit_client_protos::FunctionInfo function_info, orbit_code_report::DisassemblyReport report,
+    RetrieveModuleWithDebugInfoCallback retrieve_module_with_debug_info, QObject* parent)
+    : QObject{parent},
+      dialog_{dialog},
+      function_info_{std::move(function_info)},
+      report_{std::move(report)},
+      retrieve_module_with_debug_info_{std::move(retrieve_module_with_debug_info)} {
+  QObject::connect(dialog_, &orbit_code_viewer::Dialog::StatusMessageButtonClicked, this,
+                   &AnnotatingSourceCodeHandler::DialogActionButtonClicked);
+
+  dialog_->SetStatusMessage("Loading source location information", std::nullopt);
+
+  retrieve_module_with_debug_info_(function_info_.module_path(), function_info_.module_build_id())
+      .Then(main_thread_executor_.get(),
+            [this](const ErrorMessageOr<std::filesystem::path>& local_file_path_or_error) {
+              HandleDebugInfo(local_file_path_or_error);
+            });
+}
+
+void AnnotatingSourceCodeHandler::DialogActionButtonClicked() {
+  switch (awaited_button_action_) {
+    case ButtonAction::kNone:
+      break;
+    case ButtonAction::kChooseFile:
+      ChooseFile();
+      break;
+    case ButtonAction::kAddAnnotations:
+      HandleAnnotations();
+      break;
+  }
+}
+
+void AnnotatingSourceCodeHandler::HandleDebugInfo(
+    const ErrorMessageOr<std::filesystem::path>& local_file_path_or_error) {
```

Comment:
Personal preference, I'd split this method into a few smaller methods.

Reply:
Done. PTAL!

Reply:
Thanks, I like the update a lot! This provides a super clear picture of the flow without actually reading the details.

One nitput, and this may be controversial: I would pass the input (e.g. `elf_file_`) required in each of the steps as an argument to make it explicit, even if it it stored in a member variable. Right now, you have implicit input and output locations, which makes it less readable.

But feel free to ignore.

Reply:
I completely agree. This is why I actually prefer the previous solution. :sweat_smile: 

But I could make these functions free functions and pass in all in- and outputs. WDYT?

Reply:
Would be an option, that would also minimize the header file. I don't think that having the output implicit is a big issue though - as for the input, this prevents others that may extend the class to just call the function with wrong preconditions and then try to track down the bug.

Whatever you choose :-)

Reply:
So I tried this. But it doesn't realy work. It makes a whole mess out of everything. So I tend to refrain from making them free functions and leaving them as non-const member functions with implicit state changes.

Reply:
SGTM.